### PR TITLE
added _is_homogeneous property

### LINKED ISSUE #7067: [FEA] Internal python _is_homogeneous property
**Is your feature request related to a problem? Please describe.**
It might be useful to have a singular clean and performant way to check if all the columns of a dataframe are of the same dtype, such as a `DataFrame` property `_is_homogeneous`. This comes up in a lot of places, such as where we might want to dispatch to a cupy matrix implementation (Transpose, some row wise reductions I believe and some other places), as well as many places we might want to error or warn. 

**Describe the solution you'd like**
A property `cudf.DataFrame._is_homogeneous` that returns `True` if all of the datatypes of all of the columns are equal else false. This should be plumbed into the places where we currently use a loop or other means. This might be a good first issue being (hopefully) straightforward and requiring some somewhat broad exploration of the cuDF python layer. 

**Describe alternatives you've considered**
Writing a possibly slightly different loop every time

**Additional context**
Add any other context, code examples, or references to existing implementations about the feature request here.

### PR DESCRIPTION
**Pull request by shaneding**
This PR closes #7067.
This was implemented by adding the `_is_homogeneous` property to `DataFrame`. Included are appropriate test cases.

### CODE REVIEW

**Comment by brandon-b-miller** [comment]
thanks for adding this :) I realize now after having requested `DataFrame._is_homogenous` in the original issue that this property might be better suited for our `Frame` class so all our column based objects get it for free.

**Comment by shaneding** [comment]
@brandon-b-miller no worries, I can move this to `Frame` easily.

File: python/cudf/cudf/tests/test_dataframe.py [hunk 1]
```diff
@@ -8579,3 +8579,19 @@ def test_dataframe_init_from_series(data, columns, index):
...
+    [
+        ({"a": [1, 2, 3, 4], "b": [5, 6, 7, 8], "c": [1.2, 1, 2, 3]}, False),
+        ({"a": [1, 2, 3], "b": [4, 5, 6], "c": [7, 8, 9]}, True),
+        ({"a": ["a", "b", "c"], "b": [4, 5, 6], "c": [7, 8, 9]}, False),
+        ({"a": [True, False, False], "b": [False, False, True]}, True),
+        ({}, True),
+    ],
+)
+def test_is_homogeneous(data, expected):
+    actual = cudf.DataFrame(data)._is_homogeneous
```
Comment (brandon-b-miller) [inline python/cudf/cudf/tests/test_dataframe.py:8595]: Could we add a few tests just to make sure this works for multiindex objects/series/index etc? (the single columns are always true I suppose but might as well test for completeness)

File: python/cudf/cudf/tests/test_dataframe.py [hunk 2]
```diff
@@ -8579,3 +8579,19 @@ def test_dataframe_init_from_series(data, columns, index):
...
+
+
+@pytest.mark.parametrize(
+    "data,expected",
+    [
+        ({"a": [1, 2, 3, 4], "b": [5, 6, 7, 8], "c": [1.2, 1, 2, 3]}, False),
+        ({"a": [1, 2, 3], "b": [4, 5, 6], "c": [7, 8, 9]}, True),
+        ({"a": ["a", "b", "c"], "b": [4, 5, 6], "c": [7, 8, 9]}, False),
+        ({"a": [True, False, False], "b": [False, False, True]}, True),
+        ({}, True),
```
Comment (brandon-b-miller) [inline python/cudf/cudf/tests/test_dataframe.py:8591]: should we test the single column case as well?

**Review by brandon-b-miller — APPROVED** [review]

**Review by galipremsagar — APPROVED** [review]

**Comment by beckernick** [comment]
@gpucibot merge

**Comment by brandon-b-miller** [comment]
great work @shaneding !

**Pull request merged**